### PR TITLE
Dev change hero2vec model and results added 20200203

### PR DESCRIPTION
**Pull request by henryhao1991**
Hero2vec model added. 

hero2vec.py: The hero2vec model and training class are included. 

dataloader.py: hero2vec dataset class added to load the lineup information as the input for the training. 

hero2vec_results.ipynb: The jupyter notebook showing the results of the training and embedding.

### CODE REVIEW

File: hero2vec.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+#PyTorch libraries
```
Comment (TensorDuck) [inline hero2vec.py:1]: I can use some context here: is this `hero2vec` file ported over from the package? If so, should say so at the top of the file. If not, will do some review of the code.

File: dataloader.py [hunk 2]
```diff
@@ -58,42 +86,51 @@ def PadSequence(batch):
     It will be passed to DataLoader as the keyword argument of collate_fn.
     '''
     
+    #Get the lengths but don't do anything
+    lengths = [x["lengths"] for x in batch]
+    max_length = max(lengths)
+    lengths = torch.Tensor(lengths)
```
Comment (TensorDuck) [inline dataloader.py:92]: Small comment: 

Looks good! just wondering though, since you implement `__len__` for all of your data loaders, wouldn't you be able to get that with `len(x)`? Could be misunderstanding what batch is.

File: dataloader.py [hunk 3]
```diff
@@ -58,42 +86,51 @@ def PadSequence(batch):
     It will be passed to DataLoader as the keyword argument of collate_fn.
     '''
```
Comment (TensorDuck) [inline dataloader.py:87]: Definitely add a comment here about what datatype of batch is. Would help clarify what happened in the rest of the function.

File: dataloader.py [hunk 4]
```diff
@@ -58,42 +86,51 @@ def PadSequence(batch):
...
-    labels_padded = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True)
-    
-    #Get the lengths but don't do anything
-    lengths = [x["lengths"] for x in batch]
-    lengths = torch.Tensor(lengths)
+    labels_padded = []
+    for l in labels:
+        while l.shape[0] < max_length:
+            l = torch.cat((l, l[-1].view(1)))
+        labels_padded.append(l)
```
Comment (TensorDuck) [inline dataloader.py:108]: It is a little confusing to reassign a variable name you are iterating over. I would recommend making it clearer by changing the name of what you append to something more informative.

File: hero2vec.py (hunk 1 above)
Comment (henryhao1991) [reply hero2vec.py:1]: Oh, sorry. Should have been more specific about that. The hero2vec file uses the structure of an existing repo: https://github.com/ybw9000/hero2vec, but I modified the code to be fit with my own dataloader function and class. So it's probably 30-70, 70% based on the repo, and 30% I coded on my own.

File: dataloader.py (hunk 2 above)
Comment (henryhao1991) [reply dataloader.py:92]: This length is the length of each game, i.e., how long the game lasted, while the __len__ function for the dataloader returns the number of games in the dataset. Probably I should rename the variable as "game_length" to be more clear. I need to get this length simply because I want to use it to plot the win probability graph for a single game, but since the inputs are padded in the batch, it would be a bit tedious to get the time of the game ending without knowing the length of the game.

File: dataloader.py (hunk 4 above)
Comment (henryhao1991) [reply dataloader.py:108]: A question here. If I added a line like this:

for l in labels:
    single_label = l
    while single_label.shape[0] < max_length:
    ......

Is this essentially the same as the original code, since I remember the = operator only makes a shallow copy of the list.

File: hero2vec.py (hunk 1 above)
Comment (TensorDuck) [reply hero2vec.py:1]: Got it, will do some review on this file. You should add this to the docstring at the top of the file though so a reader will have context on this file not being 100% your original work.

File: dataloader.py (hunk 2 above)
Comment (TensorDuck) [reply dataloader.py:92]: Yeah, I think calling this game_length would add context. No need to modify any further.

File: dataloader.py (hunk 4 above)
Comment (TensorDuck) [reply dataloader.py:108]: Oh, I should have been more specific, my bad. My thinking was something even simpler:
```
    for l in labels:
        while l.shape[0] < max_length:
            single_label = torch.cat((l, l[-1].view(1)))
        labels_padded.append(single_label)
```

I'm guessing `torch.cat` makes a deep copy of the list right?

Comment (henryhao1991) [reply dataloader.py:108]: Yes, but here I need to keep padding the last line until it reaches the max length in the batch. I actually just realized `l` should be a PyTorch tensor. So probably something like:
```
    for l in labels:
        single_label = l.clone()
        while single_label.shape[0] < max_length:
            single_label = torch.cat((single_label, l[-1].view(1)))
        labels_padded.append(single_label)
```

Then in the end, `labels_padded` should be a list of length `batch_size`, with each element being a tensor of shape `(max_length, 1)`, where `max_length` is the maximum length of the game in the batch.

Comment (TensorDuck) [reply dataloader.py:108]: Yeah, your latest suggestion sounds like the right one to use for now. Good clarification.

File: dataloader.py [hunk 5]
```diff
@@ -50,6 +50,34 @@ def __getitem__(self, ind):
...
+        Returns:
+            dictionary:
+                        "context": torch.LongTensor, size = (4,). The 4 context heroes within the team.
+                        "target": int. The target hero
+        '''
+        game_ind, hero_ind = divmod(ind,5)
+        context_heroes = np.concatenate((self.lineup[game_ind, :hero_ind], self.lineup[game_ind, hero_ind+1:]))
+        target = self.lineup[game_ind, hero_ind]
+
+        return {"context":torch.LongTensor(context_heroes), "target":int(target)}
```
Comment (TensorDuck) [inline dataloader.py:80]: Well written class!

This isn't necessary for this PR, but make a note to yourself to go back and add docstrings wherever you can after week 4 is over.

File: hero2vec.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        '''
+        Inputs:
+            inputs: torch.LongTensor, size=(4,)
+        Returns:
+            out: torch.LongTensor, size=(1,)
+        '''
+        embedding = self.hero_embedding(inputs).sum(dim=1)
+        out = self.embedding_to_target(embedding)
+        
+        return out
```
Comment (TensorDuck) [inline hero2vec.py:47]: just do:

`return self.embedding_to_target(embedding)`

File: hero2vec.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        '''
+        pass
+    
+    def forward(self, inputs):
+        '''
+        Inputs:
+            inputs: torch.LongTensor, size=(4,)
+        Returns:
+            out: torch.LongTensor, size=(1,)
+        '''
```
Comment (TensorDuck) [inline hero2vec.py:43]: Understandably this is just copied over from a separate repo. But I think this area could use some more docstrings. 

Recommend:
* Add docstring about what the goal of Hero2Vec is doing.
* What the forward return should be (i.e. an embedding of a Dota Hero into some N-dimensional space).

Comment (TensorDuck) [reply hero2vec.py:43]: Oh, again: you can save that for later.

File: hero2vec.py (hunk 1 above)
Comment (TensorDuck) [reply hero2vec.py:1]: Small teachable moment:

This is where more atomic commits would've been helpful. i.e.:
```
Commit 1 - You uploading the hero2vec.py module
Commit 2 - Change you made to hero2vec.py
Commit 3 - ...
```

Not a big deal now, but just something to note as a best practice in the future. Let's talk about how best to document your work and someone else's work at our next discussion.

File: model.py [hunk 8]
```diff
@@ -59,21 +62,21 @@ def __init__(self, input_dim, hidden_dim, output_dim=1, batch_size=10, device=to
...
         Inputs:
-        inputs: torch.Tensor, size=(batch_size, input_dim, L). L is the length of the longest sequence in the batch.
-        Outputs:
-        torch.Tensor, size=(batch_size, 1, L)
+            inputs: torch.Tensor, size=(batch_size, input_dim, L). L is the length of the longest sequence in the batch.
+        Returns:
+            torch.Tensor, size=(batch_size, 1, L)
         """
         lstm_out, hidden_cell = self.lstm(inputs, self.hidden_cell)
         self.hidden_cell = tuple([hidden_.detach_() for hidden_ in hidden_cell])
```
Comment (TensorDuck) [inline model.py:82]: The changes to this file looks good 👍

File: training.py [hunk 9]
```diff
@@ -261,22 +267,21 @@ def predict(self, batch_data, threshold=0.25):
...
+        Returns:
+            predictions: torch.Tensor, size=(batch_size, L). Predictions of the probability that radiant will win. batch_size should only be 1. L is the length of the game.
         """
         with torch.no_grad():
             dim = batch_data["features"].shape
             
             #If batch_size != 1
             if len(dim) == 3:
-                self.model.hidden_cell = self.model.init_hidden()
+                self.model.hidden_cell = self.model.init_hidden(self.model.batch_size)
```
Comment (TensorDuck) [inline training.py:279]: Overall, file changes look good, mostly changes to document formatting.

Noticed this at the end though, you have this repeated pattern `self.model.init_hidden(self.model.batch_size)` throughout your code. When that happens, one thing to consider is writing for your class:

```
def init_hidden():
    self.model.hidden_cell = self.model.init_hidden(self.model.batch_size)
```

One of the motivations for this is that: you're essentially writing a wrapper class for a pytorch neural network. It is helpful sometimes to isolate a lot of the repeated interactions with the pytorch package in a single function. Since it's just one thing right now, I'll say not a necessary change. But, if you see anything else that's a repeated interaction that needs to be initialized for every run, see about consolidating all of them into a single function.

**Comment by TensorDuck** [comment]
Overall a pretty good PR. 

Only major issue would be properly documenting what is and isn't your work. Especially important since companies will be reviewing your code. Happy to discuss live sometime :D

File: training.py (hunk 9 above)
Comment (henryhao1991) [reply training.py:279]: Yeah, I was using that. But then I realized, since I need to pass in a single input to generate the graph for a single game, I have to change the `self.model.batch_size` to do that. Then later on, if I want to do something with the model, I need to change the `self.model.batch_size` back. This is the reason I decided to keep it this way, using a argument for the `init_hidden` function. Probably there is a better solution though.

**Comment by henryhao1991** [comment]
Thanks for all the feedbacks! I will just merge this branch now, and make new changes in a new branch.
